Libraries/xcexecution/Tests/test_SimpleExecutor.cpp

```diff
@@ -0,0 +1,145 @@
+/**
+ Copyright (c) 2015-present, Facebook, Inc.
+ All rights reserved.
+
+ This source code is licensed under the BSD-style license found in the
+ LICENSE file in the root directory of this source tree. An additional grant
+ of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#include <gtest/gtest.h>
+#include <xcexecution/SimpleExecutor.h>
+#include <xcformatter/NullFormatter.h>
+#include <pbxbuild/Tool/Invocation.h>
+#include <builtin/Driver.h>
+#include <builtin/Registry.h>
+#include <process/MemoryContext.h>
+#include <process/MemoryLauncher.h>
+#include <libutil/MemoryFilesystem.h>
+
+using xcexecution::SimpleExecutor;
+using libutil::Filesystem;
+using libutil::MemoryFilesystem;
+
+class Driver : public builtin::Driver {
+public:
+    using Impl = std::function<int(process::Context const *processContext, Filesystem *filesystem)>;
+
+private:
+    std::string _name;
+    Impl        _impl;
+
+public:
+    Driver(std::string const &name, Impl const &impl) :
+        _name(name),
+        _impl(impl)
+    {
+    }
+
+public:
+    virtual std::string name()
+    { return _name; }
+
+public:
+    virtual int run(process::Context const *processContext, Filesystem *filesystem)
+    { return _impl(processContext, filesystem); }
+};
+
+TEST(SimpleExecutor, PropagateToolResult)
+{
+    /* Create in-memory execution environment. */
+    auto filesystem = MemoryFilesystem({
+        MemoryFilesystem::Entry::File("fail-tool", std::vector<uint8_t>()),
+        MemoryFilesystem::Entry::File("success-tool", std::vector<uint8_t>()),
+    });
+
+    auto launcher = process::MemoryLauncher({
```

Comment:
Seems like the lowest hanging fruit for your test setup is to better unify the fake filesystem and the launcher.  Honestly, given the launcher's entries, you could create the filesystem.  Perhaps you can create a test helper function that iterates through a launcher's behavior and creates the filesystem underneath it?